### ISSUE TITLE
Can't close a Modal using JS

### ISSUE THREAD

**Issue by caub** [twbs/bootstrap#33727]
- OS: linux
- browser: Chrome latest
- test-case: https://jsfiddle.net/caub/x6t2ovqm/

**Comment by alpadev** [comment]
Thanks for reporting this problem. You're indeed correct with your expectation that this should be handled by our code.

As a workaround until it's fixed you may want to use:
```js
const closeWithJS = document.getElementById('closeWithJS');
const openWithJS = document.getElementById('openWithJS');
const modalEl = document.getElementById('exampleModal');

let modal;

closeWithJS.onclick = () => {
  const modal = bootstrap.Modal.getInstance(modalEl);
  modal.hide();
  modalEl.addEventListener('hidden.bs.modal', () => {
    modal.dispose();
  }, {once:true});
};

openWithJS.onclick = () => {
  modal = new bootstrap.Modal(modalEl);
  modal.show();
};
```

cc @GeoSot

**Comment by GeoSot** [comment]
It applies to all components. :fearful: 
But yes, we need to see this

**Comment by rohit2sharma95** [comment]
So do you guys want the hide method to work synchronously? :worried:
Since the code that you wrote is fine IMO @alpadev
```js
const modal = bootstrap.Modal.getInstance(modalEl);
modal.hide();
modalEl.addEventListener('hidden.bs.modal', () => {
  modal.dispose();
}, { once:true });
```

/CC @GeoSot

**Comment by alpadev** [comment]
> So do you guys want the hide method to work synchronously? 😟
> Since the code that you wrote is fine IMO @alpadev
> 
> ```js
> const modal = bootstrap.Modal.getInstance(modalEl);
> modal.hide();
> modalEl.addEventListener('hidden.bs.modal', () => {
>   modal.dispose();
> }, { once:true });
> ```
> 
> /CC @GeoSot

@rohit2sharma95 Hm not sure. I was thinking that it might be a bit tough to know how to do it the right way without knowing the code better. Maybe it would be enough to put an example into the docs.

**Comment by GeoSot** [comment]
> So do you guys want the hide method to work synchronously? 😟

No, for any reason. :) will brake many UX things for sure

After #33609 merge, we can discuss a holistic solution, but yet I am not sure, cause it would transform the `dispose` method, form sync to potential  async

**Comment by caub** [comment]
Should there be something like a `close` function that will wrap all the:

```js
modal.hide();
modalEl.addEventListener('hidden.bs.modal', () => {
  modal.dispose();
}, { once:true });
```

**Comment by GeoSot** [comment]
>Should there be something like a close function that will wrap all the:

I am not sure if this is a proper solution :thinking:

**Comment by caub** [comment]
I think we can close? updated to bootstrap 5.0.1 https://jsfiddle.net/caub/x6t2ovqm/3/ and it works

**Comment by caub** [comment]
Sorry, no, the "Close with JS" button is still not working

**Comment by alpadev** [comment]
@caub glad if this works for you.

I'd prefer to leave this open tho, so that we can track this and maybe at some point come up with an easier solution.

**Comment by dong94vk** [comment]
I see when creating modal there are 2 div.modal-backdrop.show created on DOM. but when hiding the modal with JS it only removes 1 div.modal-backdrop.show so it still has the backdrop behind. Maybe this is the problem

**Comment by dong94vk** [comment]
> Thanks for reporting this problem. You're indeed correct with your expectation that this should be handled by our code.
> 
> As a workaround until it's fixed you may want to use:
> 
> ```js
> const closeWithJS = document.getElementById('closeWithJS');
> const openWithJS = document.getElementById('openWithJS');
> const modalEl = document.getElementById('exampleModal');
> 
> closeWithJS.onclick = () => {
>   const modal = bootstrap.Modal.getInstance(modalEl);
>   modal.hide();
>   modalEl.addEventListener('hidden.bs.modal', () => {
>     modal.dispose();
>   }, {once:true});
> };
> 
> openWithJS.onclick = () => {
>   const modal = new bootstrap.Modal(modalEl);
>   modal.show();
> };
> ```
> 
> cc @GeoSot

I'm using it in react but it's not work :(

**Comment by miguellira** [comment]
Ran into this issue when looking into using bootstrap 5 with Angular. Long story short: 
```
Do NOT mix bootstrap's data attributes with javascript calls. 
```
If you need a modal to display content with no other user interaction except dismissing the modal then the data attribute approach is simple and works fine. However, if you have user interaction in your modal then remove all `data-bs-****` attributes and call the `hide()` on the `Modal` type.

The following `Angular` code works as expected.

**app.component.ts**
```typescript
import { AfterViewInit, Component, ElementRef, ViewChild } from '@angular/core';
import { Modal } from 'bootstrap';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements AfterViewInit {

  @ViewChild('exampleModal') exampleModalRef!: ElementRef;

  exampleModal!: Modal;

  ngAfterViewInit(): void {
    this.exampleModal = new Modal(this.exampleModalRef.nativeElement);
  }

  showModal(): void {
    this.exampleModal.show();
  }

  hideModal(): void {
    this.exampleModal.hide();
  }

  saveChanges(e: any): void {
    console.log('save changes clicked');
    this.hideModal();
  }
}
```
**app.component.html**
```html
<div class="container">
  <!-- Button trigger modal -->
  <button type="button" class="btn btn-primary" (click)="showModal()">
    Launch demo modal
  </button>

  <!-- Modal -->
  <div class="modal fade" #exampleModal id="exampleModal" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title" id="exampleModalLabel">Modal title</h5>
          <button type="button" class="btn-close" aria-label="Close"></button>
        </div>
        <div class="modal-body">
          ...
        </div>
        <div class="modal-footer">
          <button type="button" class="btn btn-secondary" (click)="hideModal()">Close</button>
          <button type="button" class="btn btn-primary" (click)="saveChanges($event)">Save changes</button>
        </div>
      </div>
    </div>
  </div>
</div>
```

**Comment by miguellira** [comment]
@bluesurfer Does your template contain any data

**Comment by bluesurfer** [comment]
@miguellira mmm actually no.. I am using javascript to populate the content of the modal (see https://getbootstrap.com/docs/5.2/components/modal/#varying-modal-content)

**Comment by miguellira** [comment]
What about the modal itself? Maybe top right close button? Check the emitted DOM. TBH it's been months since I posted in this thread and as I mentioned it all worked fine once I removed all `data-bs` attributes.

**Comment by bluesurfer** [comment]
@miguellira wait are you telling me not to use **any** `data-bs` attribute not even for popover or tooltip? I am simply not using any `data-bs-modal`

**Comment by bluesurfer** [comment]
@miguellira aah I solve it.. For me the problem are not the data attributes but when I was creating/updating the modal through Javascript I keep calling `new Modal`. That makes my code was "ill" at the very beginning 😄  

Using `Modal.getOrCreateInstance` at modal creation/update and then `Modal.getInstance` to close it at form submit (in my case) did the trick: the `hide()` method works as expected.

**Comment by dinoop-caseking** [comment]
`modal` object should be initialized out of the click event, in this way it worked for me. 

```
const modalEle: HTMLElement = document.querySelector('#js-modal');
        const modal = new Modal(modalEle, {});

        document.querySelector('.close-buttom').addEventListener('click', (evt: Event) => {
            if (modalEle){
                modal.hide();
            }
        });
```

**Comment by Maritims** [comment]
> Ran into this issue when looking into using bootstrap 5 with Angular. Long story short:
> 
> ```
> Do NOT mix bootstrap's data attributes with javascript calls. 
> ```
> 

Thanks! Seems like showing and hiding a modal using Bootstrap 5 and Vue 3 should be handled explicitly through the JS methods and entirely without the data-bs- attributes. I don't know why though.

**Comment by sven-straubinger** [comment]
> > Ran into this issue when looking into using bootstrap 5 with Angular. Long story short:
> > ```
> > Do NOT mix bootstrap's data attributes with javascript calls. 
> > ```
> 
> Thanks! Seems like showing and hiding a modal using Bootstrap 5 and Vue 3 should be handled explicitly through the JS methods and entirely without the data-bs- attributes. I don't know why though.

I can confirm this fixed the issue for me in Bootstrap 5.3.8. I removed all data- attributes and handled the modal entirely with JavaScript for initialization, showing, and hiding. Commenting this here in case someone is still looking for a solution in 2025+.

**Comment by mdo** [comment]
v6-dev rebuilds modals on the native `<dialog>` element, so there's no separate backdrop node to leave behind. Closing as resolved there — please open a new issue if this persists for you in v6-dev.
